### ISSUE TITLE
Fail to build on clang-8

### ISSUE THREAD

**Issue by sneumann** [sneumann/mzR#190]
We've got a report from Prof Brian Ripley:
```
These packages fail to install with clang 8.0.0rc3, which is supposed to 
be near-final (it is overdue for release).  (AFAIR they did install with 
rc1.)  In both cases the problems are with Boost.

mzR 2.17.1

In file included from pwiz/data/identdata/Serializer_pepXML.cpp:25:
In file included from pwiz/data/identdata/Serializer_pepXML.hpp:27:
In file included from pwiz/data/identdata/IdentData.hpp:29:
In file included from ./pwiz/utility/misc/Exception.hpp:29:
In file included from /usr/local/clang8/bin/../include/c++/v1/string:505:
In file included from 
/usr/local/clang8/bin/../include/c++/v1/string_view:176:
In file included from /usr/local/clang8/bin/../include/c++/v1/__string:57:
/usr/local/clang8/bin/../include/c++/v1/algorithm:2428:5: error: 
static_assert
       failed due to requirement
 
'__is_forward_iterator<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_specificity, std::__1::__wrap_iter<const
       boost::shared_ptr<pwiz::identdata::Enzyme> *>,
       boost::iterators::use_default, boost::iterators::use_default> 
 >::value'
       "std::min_element requires a ForwardIterator"
     static_assert(__is_forward_iterator<_ForwardIterator>::value,
     ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/clang8/bin/../include/c++/v1/algorithm:2445:19: note: in
       instantiation of function template specialization
 
'std::__1::min_element<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_specificity, std::__1::__wrap_iter<const
       boost::shared_ptr<pwiz::identdata::Enzyme> *>,
       boost::iterators::use_default, boost::iterators::use_default>,
       std::__1::__less<int, int> >' requested here
     return _VSTD::min_element(__first, __last,
                   ^
./boost/boost/range/algorithm/min_element.hpp:44:17: note: in 
instantiation of
       function template specialization
 
'std::__1::min_element<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_specificity, std::__1::__wrap_iter<const
       boost::shared_ptr<pwiz::identdata::Enzyme> *>,
       boost::iterators::use_default, boost::iterators::use_default> >' 
requested
       here
     return std::min_element(boost::begin(rng), boost::end(rng));



pwiz/data/identdata/Serializer_pepXML.cpp:418:41: note: in instantiation of
       function template specialization
 
'boost::range::min_element<boost::range_detail::transformed_range<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_specificity, const
       std::__1::vector<boost::shared_ptr<pwiz::identdata::Enzyme>,
       std::__1::allocator<boost::shared_ptr<pwiz::identdata::Enzyme> > 
 > > >'
       requested here
     int minSpecificity = *boost::range::min_element(sip.enzymes.enzymes ...
                                         ^
In file included from pwiz/data/identdata/Serializer_pepXML.cpp:25:
In file included from pwiz/data/identdata/Serializer_pepXML.hpp:27:
In file included from pwiz/data/identdata/IdentData.hpp:29:
In file included from ./pwiz/utility/misc/Exception.hpp:29:
In file included from /usr/local/clang8/bin/../include/c++/v1/string:505:
In file included from 
/usr/local/clang8/bin/../include/c++/v1/string_view:176:
In file included from /usr/local/clang8/bin/../include/c++/v1/__string:57:
/usr/local/clang8/bin/../include/c++/v1/algorithm:2494:5: error: 
static_assert
       failed due to requirement
 
'__is_forward_iterator<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_missedCleavages, std::__1::__wrap_iter<const
       boost::shared_ptr<pwiz::identdata::Enzyme> *>,
       boost::iterators::use_default, boost::iterators::use_default> 
 >::value'
       "std::max_element requires a ForwardIterator"
     static_assert(__is_forward_iterator<_ForwardIterator>::value,
     ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/clang8/bin/../include/c++/v1/algorithm:2512:19: note: in
       instantiation of function template specialization
 
'std::__1::max_element<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_missedCleavages, std::__1::__wrap_iter<const
       boost::shared_ptr<pwiz::identdata::Enzyme> *>,
       boost::iterators::use_default, boost::iterators::use_default>,
       std::__1::__less<int, int> >' requested here
     return _VSTD::max_element(__first, __last,
                   ^
./boost/boost/range/algorithm/max_element.hpp:44:17: note: in 
instantiation of
       function template specialization
 
'std::__1::max_element<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
       namespace)::EnzymePtr_missedCleavages, std::__1::__wrap_iter<const
       boost::shared_ptr<pwiz::identdata::Enzyme> *>,
       boost::iterators::use_default, boost::iterators::use_default> >' 
requested
       here
     return std::max_element(boost::begin(rng), boost::end(rng));
...

```

**Comment by sneumann** [comment]
I am using packages from https://apt.llvm.org/ I can confirm mzR-2.17.1 builds successfully with 
`clang version 8.0.0-svn354892-1~exp1~20190226195658.47 (branches/release_80)`
which seems to be a bit newer than rc2, but not yet at rc3. I'll retry once we have updated packages.
yours, Steffen

**Comment by sneumann** [comment]
Hi, just tested with `clang version 8.0.0-svn355584-1~exp1~20190307092823.52 (branches/release_80)` 
which seems to be newer that `llvm-toolchain-8 (1:8~+rc2-1~exp2)` but not yet at `clang 8.0.0rc3` that has the failure reported above Still waiting for rc3 binaries to test. Yours, Steffen

**Comment by sneumann** [comment]
ok on `clang version 8.0.0-svn355916-1~exp1~20190312132827.53 (branches/release_80)`

**Comment by sneumann** [comment]
I still can't reproduce the failure. The Dockerfile below is based 
on https://hub.docker.com/r/rocker/r-base/ , and `Makevars` switches to 
`clang version 8.0.0-1 (tags/RELEASE_800/final)`

Help needed to reproduce. Yours, Steffen

```
R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux buster/sid

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mzR_2.17.2 Rcpp_1.0.1
```


[Dockerfile.txt](https://github.com/sneumann/mzR/files/2996151/Dockerfile.txt)
[Makevars.txt](https://github.com/sneumann/mzR/files/2996152/Makevars.txt)

**Comment by jotsetung** [comment]
On macOS I get the following error (R-3.6 devel compiled with llvm 8.0 (release)):
```
In file included from pwiz/data/identdata/Serializer_pepXML.cpp:25:
In file included from pwiz/data/identdata/Serializer_pepXML.hpp:27:
In file included from pwiz/data/identdata/IdentData.hpp:29:
In file included from ./pwiz/utility/misc/Exception.hpp:29:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/string:505:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/string_view:176:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/__string:57:
/usr/local/opt/llvm/bin/../include/c++/v1/algorithm:2494:5: error: static_assert failed due to requirement
      '__is_forward_iterator<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
      namespace)::EnzymePtr_missedCleavages, std::__1::__wrap_iter<const boost::shared_ptr<pwiz::identdata::Enzyme> *>,
      boost::iterators::use_default, boost::iterators::use_default> >::value' "std::max_element requires a ForwardIterator"
    static_assert(__is_forward_iterator<_ForwardIterator>::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/opt/llvm/bin/../include/c++/v1/algorithm:2512:19: note: in instantiation of function template specialization
      'std::__1::max_element<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
      namespace)::EnzymePtr_missedCleavages, std::__1::__wrap_iter<const boost::shared_ptr<pwiz::identdata::Enzyme> *>,
      boost::iterators::use_default, boost::iterators::use_default>, std::__1::__less<int, int> >' requested here
    return _VSTD::max_element(__first, __last,
                  ^
./boost/boost/range/algorithm/max_element.hpp:44:17: note: in instantiation of function template specialization
      'std::__1::max_element<boost::iterators::transform_iterator<pwiz::identdata::(anonymous
      namespace)::EnzymePtr_missedCleavages, std::__1::__wrap_iter<const boost::shared_ptr<pwiz::identdata::Enzyme> *>,
      boost::iterators::use_default, boost::iterators::use_default> >' requested here
    return std::max_element(boost::begin(rng), boost::end(rng));
                ^
pwiz/data/identdata/Serializer_pepXML.cpp:504:49: note: in instantiation of function template specialization
      'boost::range::max_element<boost::range_detail::transformed_range<pwiz::identdata::(anonymous
      namespace)::EnzymePtr_missedCleavages, const std::__1::vector<boost::shared_ptr<pwiz::identdata::Enzyme>,
      std::__1::allocator<boost::shared_ptr<pwiz::identdata::Enzyme> > > > >' requested here
        int maxMissedCleavages = *boost::range::max_element(sip.enzymes.enzymes | boost::adaptors::transformed(EnzymePtr_mi...
                                                ^
2 warnings and 2 errors generated.
make: *** [pwiz/data/identdata/Serializer_pepXML.o] Error 1
ERROR: compilation failed for package ‘mzR’
* removing ‘/Users/jo/R/llvm8/R-3.6-2019-03-25-Bioc3.9-devel/lib/R/library/mzR’


```

**Comment by sneumann** [comment]
Thanks to Kurt Hornik the issue can now also be reproduced with the updated Dockerfile and Makevars. 
```
clang++-8 -stdlib=libc++ -std=gnu++11 -I"/usr/share/R/include" -DNDEBUG -I./boost_aux/ -I./boost/ -I. -D_LARGEFILE_SOURCE -DHAVE_PWIZ_MZML_LIB -D_NODEBUG    -I"/usr/local/lib/R/site-library/Rcpp/include" -I"/usr/local/lib/R/site-library/zlibbioc/include" -I"/usr/local/lib/R/site-library/Rhdf5lib/include"    -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-3.5.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g -c pwiz/data/identdata/Serializer_pepXML.cpp -o pwiz/data/identdata/Serializer_pepXML.o
```
Yours, Steffen
[Dockerfile.txt](https://github.com/sneumann/mzR/files/3004216/Dockerfile.txt)
[Makevars.txt](https://github.com/sneumann/mzR/files/3004217/Makevars.txt)

**Comment by sneumann** [comment]
The compile error looks similar to this one: 
https://www.mail-archive.com/llvm-bugs@lists.llvm.org/msg30845.html
and happens in https://github.com/sneumann/mzR/blob/master/src/pwiz/data/identdata/Serializer_pepXML.cpp#L504
```
int maxMissedCleavages = *boost::range::max_element(sip.enzymes.enzymes | boost::adaptors::transformed(EnzymePtr_missedCleavages()));`
```
which seems to apply the function `EnzymePtr_missedCleavages()`
defined in https://github.com/sneumann/mzR/blob/master/src/pwiz/data/identdata/Serializer_pepXML.cpp#L397
to all elements in the `SpectrumIdentificationProtocol sip`.

N.B.: In the branch with the updated pwiz+boost there is no error. 
The fix could be either in the `boost::adaptors::*` world, OR in pwiz. 

Yours, Steffen

**Comment by sneumann** [comment]
Maybe @David-J-R is a capable C++ hacker here ... Yours, Steffen

**Comment by sneumann** [comment]
Thanks to K-Ballo on IRC #boost I was pointed to this commit:
https://github.com/boostorg/range/commit/adcb071dc6b06e69aa4ddcc080d23695a5da378f
Here we include Boost-1.67.0, which is affected (and possibly fixed) by that commit. Will try to replace 
just those boost headers in-place. Yours, Steffen

**Comment by sneumann** [comment]
We now get much further, up to a link error. Note that there is now a branch 
https://github.com/sneumann/mzR/tree/fix/issue190 to work on the issue.
```
Error: package or namespace load failed for ‘mzR’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/usr/local/lib/R/site-library/mzR/libs/mzR.so':
  /usr/local/lib/R/site-library/mzR/libs/mzR.so: undefined symbol: _ZNK2H510H5Location13createDataSetERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEERKNS_8DataTypeERKNS_9DataSpaceERKNS_17DSetCreatPropListERKNS_15DSetAccPropListERKNS_17LinkCreatPropListE
Error: loading failed
```
[Dockerfile.txt](https://github.com/sneumann/mzR/files/3031522/Dockerfile.txt)
[Makevars.txt](https://github.com/sneumann/mzR/files/3031524/Makevars.txt)

**Comment by David-J-R** [comment]
I think the error occurs in file pwiz/data/msdata/mz5/Connection_mz5.cpp at line 119:

```
DataSet dataset = file_->createDataSet(config_.getNameFor(v),
            config_.getDataTypeFor(v), dataSpace, cparms);
```

which is called in the same file in line 135 and 253. 
The problem could be that the H5Cpp.h isn't included in that file.

**Comment by sneumann** [comment]
Turns out that the fix in 1423cb408de623c0cc89e3e9d95d22b67f879232 already solves this issue,
and that my testing environment was flawed. The Dockerfile had added the clang Makevars 
before compiling mzR, but after compiling Rhdf5lib. Mixing compilers when linking the results together 
is not a good thing (tm). Updated Dockerfile and Makevars here. @jorainer (or someone with a mac) 
could you R CMD check on macOS before I merge into master ? Yours, Steffen

[Dockerfile.txt](https://github.com/sneumann/mzR/files/3190000/Dockerfile.txt)
[Makevars.txt](https://github.com/sneumann/mzR/files/3190002/Makevars.txt)

**Comment by jorainer** [comment]
I will. Just to be sure: I'll check `mzR` branch *fix/issue190* on R-3.6.0 built with clang 8, right?

**Comment by jorainer** [comment]
`R CMD check` of *fix/issue190* successfully ran through on an R-3.6.0 compiled with llvm 8 with a single WARNING:

```
* checking whether package ‘mzR’ can be installed ... WARNING
Found the following significant warnings:
  pwiz/utility/misc/TabReader.cpp:44:22: warning: assigning field to itself [-Wself-assign-field]
Found the following additional notes/warnings:
  pwiz/data/identdata/Serializer_pepXML.cpp:1545:5: warning: control may reach end of non-void function [-Wreturn-type]
See ‘/Users/jo/Projects/git/sneumann/mzR.Rcheck/00install.out’ for details.

```

So, seems to be OK on macOS
